src/lib/cpuset.lua

```diff
@@ -19,35 +19,78 @@ do
    end
 end
 
+local function parse_cpulist (cpus)
+   local ret = {}
+   for range in cpus:split(',') do
+      local lo, hi = range:match("^%s*([^%-]*)%s*-%s*([^%-%s]*)%s*$")
+      if lo == nil then lo = range:match("^%s*([^%-]*)%s*$") end
+      assert(lo ~= nil, 'invalid range: '..range)
+      lo = assert(tonumber(lo), 'invalid range begin: '..lo)
+      assert(lo == math.floor(lo), 'invalid range begin: '..lo)
+      if hi ~= nil then
+         hi = assert(tonumber(hi), 'invalid range end: '..hi)
+         assert(hi == math.floor(hi), 'invalid range end: '..hi)
+         assert(lo < hi, 'invalid range: '..range)
+      else
+         hi = lo
+      end
+      for cpu=lo,hi do table.insert(ret, cpu) end
+   end
+   return ret
+end
+
+local function parse_cpulist_from_file (path)
+   local fd = assert(io.open(path))
+   if not fd then return {} end
+   local ret = parse_cpulist(fd:read("*all"))
+   fd:close()
+   return ret
+end
+
+local function available_cpus (node)
+   local function set (t)
+      local ret = {}
+      for _,v in pairs(t) do ret[tostring(v)] = true end
+      return ret
+   end
+   local function cpus_in_node (node)
+      local node_path = '/sys/devices/system/node/node'..node
+      return parse_cpulist_from_file(node_path..'/cpulist')
+   end
+   local function isolated_cpus ()
+      return parse_cpulist_from_file('/sys/devices/system/cpu/isolated')
```

Comment:
Clever!  I had no idea this file was a thing.

Two requests:

  (1) Can we make a method to subtract cpulists ?  That way avail = subtract(cpus_in_node, isolated_cpus).

  (2) Can we set affinity in `bind_to_numa_node` to the interserction of the current CPU affinity with all non-isolated CPUs on a node, not just the first one?  That will give Linux a bit more freedom to optimize these non-performance-critical processes, while also preserving the user's ability to do `taskset -c N`.